### PR TITLE
rgw_file: add lock op support

### PR DESCRIPTION
**Pull request by guihecheng**
+ rgw_file flock api
  - setlk
  - getlk
+ unittest
  - librgw_file_locks 

for now, flocks are managed in memory only, we shall consolidate them later.

### CODE REVIEW

File: src/include/rados/rgw_file.h [hunk 1]
```diff
@@ -84,6 +84,21 @@ struct rgw_statvfs {
...
 };
 
+enum rgw_fl_type {
+  RGW_FS_TYPE_FL_RDLOCK = 0,
+  RGW_FS_TYPE_FL_WRLOCK,
+  RGW_FS_TYPE_FL_UNLOCK,
+};
+
+struct rgw_flock {
+    uint64_t start;
```
Comment (mattbenjamin) [inline src/include/rados/rgw_file.h]: off, len

File: src/include/rados/rgw_file.h [hunk 2]
```diff
@@ -84,6 +84,21 @@ struct rgw_statvfs {
...
+    uint64_t length;
+    uint64_t client;
```
Comment (mattbenjamin) [inline src/include/rados/rgw_file.h]: we might need a different data type for client

File: src/include/rados/rgw_file.h [hunk 3]
```diff
@@ -337,6 +352,19 @@ int rgw_fsync(struct rgw_fs *rgw_fs, struct rgw_file_handle *fh,
...
 	       uint64_t offset, uint64_t length, uint32_t flags);
 
+
+/*
+   NFS lock operation
+*/
+
+#define RGW_LOCK_FLAG_NONE          0x0000
+
+int rgw_setlk(struct rgw_fs *rgw_fs, struct rgw_file_handle *fh,
```
Comment (mattbenjamin) [inline src/include/rados/rgw_file.h]: ok

File: src/rgw/rgw_file.cc [hunk 4]
```diff
@@ -1355,6 +1357,250 @@ namespace rgw {
     return op_ret;
   } /* exec_continue */
 
+  int RGWFileHandle::insert_flock(flock &target, uint32_t flags)
+  {
+    list<multimap<uint64_t, flock>::iterator> overlaps, neighbors;
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.cc]: we'd like to support efficient selection of client(s) preferred or selected to get a lock when blocking locks are released;  freebsd has a dependency graph for this, see https://github.com/freebsd/freebsd/blob/master/sys/kern/kern_lockf.c

**Comment by mattbenjamin** [comment]
@guihecheng looks great so far

File: src/rgw/rgw_file.cc (hunk 4 above)
Comment (guihecheng) [inline src/rgw/rgw_file.cc]: @mattbenjamin oh, thanks for your information, I shall check this.

**Comment by tchaikov** [comment]
@guihecheng ping?

**Comment by guihecheng** [comment]
@tchaikov oh, sorry for being idle for long, I've got a draft based on the model that matt suggested, it needs further tests, I'll send an update in 2 weeks.

**Comment by tchaikov** [comment]
@guihecheng thanks!

**Comment by guihecheng** [comment]
Sorry for holding this for long, and here I have a version that passes my simple tests.
This version adopts the model that recommended by @mattbenjamin , with blocking lock support(still no async notify for now) and deadlock detection.
Maybe you're still interest on this one, @mattbenjamin ? All comments appreciated :)

**Comment by mattbenjamin** [comment]
@guihecheng needs rebase

**Comment by guihecheng** [comment]
@mattbenjamin very glad to receive comments from you, matt. This pr was written only as a preview for the flock feature when I frist drafted it and thanks for your comments, I'll check it later.
Richer IO models are of the highest priority indeed, as we discussed before, but it seems that the community shows very little interest on RGW-NFS itself, for now?
By the way, I've shifted from one company to another for the last several months and paid most of my effort to some glusterfs stuff. But for now, I think I shall have some time to continue cephing on!

**Pull request closed without merging**